les/client_handler.go

```diff
@@ -45,6 +46,47 @@ type clientHandler struct {
 	closeCh  chan struct{}
 	wg       sync.WaitGroup // WaitGroup used to track all connected peers.
 	syncDone func()         // Test hooks when syncing is done.
+
+	//@rayyuan
+	gatewayFeeCache *gatewayFeeCache
+}
+
+type gatewayFeeEtherbase struct {
+	GatewayFee *big.Int
+	Etherbase  common.Address
+}
+
+func newGatewayFeeEtherbase(gatewayFee *big.Int, etherbase common.Address) *gatewayFeeEtherbase {
+	ret := &gatewayFeeEtherbase{
+		GatewayFee: gatewayFee,
+		Etherbase:  etherbase,
+	}
+	return ret
+}
+
+type gatewayFeeCache struct {
+	mux           *sync.RWMutex
+	gatewayFeeMap map[string]*gatewayFeeEtherbase
+}
+
+func newGatewayFeeCache() *gatewayFeeCache {
+	cache := &gatewayFeeCache{
+		mux:           new(sync.RWMutex),
+		gatewayFeeMap: make(map[string]*gatewayFeeEtherbase),
+	}
+	return cache
+}
+
+func (c *gatewayFeeCache) update(nodeID string, val *gatewayFeeEtherbase) error {
```

Comment:
I see that updates are done under a lock, but that reads are accessed directly. It is unsafe to access 
the values without a lock, so functions will need to be added for that as well.

Reply:
for reads, would it be okay to just do h.gatewayFeeCache.mutex.Lock(), or should I have another function for it?

Reply:
You can use the same lock method. In reads are expected to be **much** more common than writes you could consider using a `RWMutex` (https://golang.org/pkg/sync/#RWMutex) instead, but I would not worry about that too much unless performance is shown to be an issue in production.